api/envoy/extensions/filters/http/local_ratelimit/v3/local_rate_limit.proto

```diff
@@ -0,0 +1,59 @@
+syntax = "proto3";
+
+package envoy.extensions.filters.http.local_ratelimit.v3;
+
+import "envoy/config/core/v3/base.proto";
+import "envoy/type/v3/http_status.proto";
+import "envoy/type/v3/token_bucket.proto";
+
+import "udpa/annotations/status.proto";
+import "udpa/annotations/versioning.proto";
+import "validate/validate.proto";
+
+option java_package = "io.envoyproxy.envoy.extensions.filters.http.local_ratelimit.v3";
+option java_outer_classname = "LocalRateLimitProto";
+option java_multiple_files = true;
+option (udpa.annotations.file_status).package_version_status = ACTIVE;
+
+// [#protodoc-title: Local Rate limit]
+// Local Rate limit :ref:`configuration overview <config_http_filters_local_rate_limit>`.
+// [#extension: envoy.filters.http.local_ratelimit]
+
+// [#next-free-field: 11]
+message LocalRateLimit {
+  option (udpa.annotations.versioning).previous_message_type =
+      "envoy.config.filter.http.local_rate_limit.v2.LocalRateLimit";
+
+  // This field allows to send a HTTP response status code to the downstream client other
+  // than 429 (TooManyRequests) when the request has been rate limited.
+  type.v3.HttpStatus status = 1;
+
+  // The token bucket configuration to use for rate limiting requests that are processed by this
+  // filter. Each request processed by the filter consumes a single token. If the token is available,
+  // the request will be allowed. If no tokens are available, the request will receive the configured
+  // rate limit status.
+  //
+  // .. note::
+  //   It's fine for the token bucket to not be set for the global configuration, but it must be set
+  //   for the per route configuration.
+  //
+  // .. note::
+  //   When using per route configuration, the bucket becomes unique to that route.
+  //
+  // .. note::
+  //   In the current implementation the token bucket's :ref:`fill_interval
+  //   <envoy_api_field_type.v3.TokenBucket.fill_interval>` must be >= 50ms to avoid too aggressive
+  //   refills.
+  type.v3.TokenBucket token_bucket = 2;
+
+  // This uniquely identifies a specific local rate limit configuration (e.g.: when using per route
+  // rate limiters). It will be used to construct the
+  // :ref:`runtime keys <config_http_filters_local_rate_limit_runtime>` that enable and enforce
+  // the corresponding local rate limiter.
+  string route_key = 3 [(validate.rules).string = {min_bytes: 1}];
+
+  // Specifies a list of HTTP headers that should be added to each response for requests that
+  // have been rate limited.
+  repeated config.core.v3.HeaderValueOption response_headers_to_add = 10
+      [(validate.rules).repeated = {max_items: 10}];
+}
```

Comment:
Did you consider making this a variation on the existing rate limit filter, with some additional config for local response? Asking since it might be the case that there is a bunch of shared config, e.g. `request_type`, `failure_mode_deny`.

Reply:
I did, but ended up concluding that there is much more in common with the network local rate limiter (than the existing HTTP global rate limiter).

@mattklein123 thoughts?

Reply:
I think originally had a similar inclination to @htuch about building this into the existing rate limit filter, but seeing this config it's simpler enough that I can buy having a separate filter for it. The only thing that would give me pause is whether someone is going to want to match on different types of things (descriptors) and whether we might be better served by constructing a descriptor as a key to a local token bucket? I'm not sure if this is worth it or not but just throwing it out there and I obviously didn't bother doing this for the network rate limit case (albeit a much simpler case). Thoughts?